client/offchain/src/api/http.rs

```diff
@@ -693,283 +693,294 @@ impl fmt::Debug for HttpWorkerRequest {
 #[cfg(test)]
 mod tests {
 	use crate::api::timestamp;
-	use super::http;
-	use futures::prelude::*;
-	use futures01::Future as _;
+	use super::{http, HttpApi};
 	use sp_core::offchain::{HttpError, HttpRequestId, HttpRequestStatus, Duration};
+	use std::{net::SocketAddr, convert::Infallible};
+	use hyper::{Request, Response, Body, service::{make_service_fn, service_fn}};
+	use tokio::runtime::Runtime;
+
+	async fn hello_world(_: Request<Body>) -> Result<Response<Body>, Infallible> {
+		Ok(Response::new(Body::from("Hello World!")))
+	}
 
 	// Returns an `HttpApi` whose worker is ran in the background, and a `SocketAddr` to an HTTP
 	// server that runs in the background as well.
-	macro_rules! build_api_server {
-		() => {{
-			let (api, worker) = http();
-			// Note: we have to use tokio because hyper still uses old futures.
-			std::thread::spawn(move || {
-				tokio::run(futures::compat::Compat::new(worker.map(|()| Ok::<(), ()>(()))))
-			});
-			let (addr_tx, addr_rx) = std::sync::mpsc::channel();
-			std::thread::spawn(move || {
-				let server = hyper::Server::bind(&"127.0.0.1:0".parse().unwrap())
-					.serve(|| {
-						hyper::service::service_fn_ok(move |_: hyper::Request<hyper::Body>| {
-							hyper::Response::new(hyper::Body::from("Hello World!"))
-						})
-					});
-				let _ = addr_tx.send(server.local_addr());
-				hyper::rt::run(server.map_err(|e| panic!("{:?}", e)));
-			});
-			(api, addr_rx.recv().unwrap())
-		}};
+	fn build_api_server() -> (HttpApi, SocketAddr) {
```

Comment:
The original reason for using background threads is so that we wouldn't depend on being polled by tokio.

Reply:
The tests seem to still be fine. Do you want. me to change it back to using background threads?

Reply:
Considering that this crosses the "runtime boundary", it makes it less easy to figure out whether we're polling using tokio, so I'd say yes.

Reply:
Tokio is annoyingly tricky, and I haven't figured out how to do this without getting a panic due to runtime stuff. I'll add a comment that should hopefully make things clearer though.

Reply:
I just removed the offchain changes because I couldn't work out how `net2` was getting pulled in and I would like to merge this PR soon. Probably caused by https://github.com/rust-lang/cargo/issues/4866.